### ISSUE TITLE
[BUG REPORT] Commit inside context manager throws RuntimeError

### ISSUE THREAD

**Issue by hhsecond** [tensorwerk/hangar-py#81]
**Describe the bug**
If we try to commit inside the context manager (before `__exit__()`), hangar throws RuntimeError saying `No changes made in the staging area. Cannot commit.`. We should allow the user to do commits inside the context manager IMO but probably with a warning about the performance hit

**Severity**

Select an option:
- [ ] Data Corruption / Loss of Any Kind
- [x] Unexpected Behavior, Exceptions or Error Thrown
- [ ] Performance Bottleneck

**To Reproduce**
```python
import numpy as np

from hangar import Repository
repo = Repository(path='myhangarrepo')
repo.init(user_name='Sherin Thomas', user_email='sherin@gmail.com', remove_old=True)

# generate data
data = []
for i in range(1000):
    data.append(np.random.rand(28, 28))
data = np.array(data)

co = repo.checkout(write=True)
data_dset = co.datasets.init_dataset('mnist_data', prototype=data[0])
co.commit('datasets init')
co.close()
co = repo.checkout(write=True)
data_dset = co.datasets['mnist_data']

with data_dset:
    for i in range(len(data)):
        sample_name = str(i)
        data_dset[sample_name] = data[i]
        co.commit('dataset curation: stage 1')  # this throws error
co.close()
```

**Expected behavior**
It should not break the program instead raise a warning about the performance hit

**Comment by hhsecond** [comment]
@lantiga I remember telling you this won't break the program instead raise an issue. I was wrong. It raise a `RuntimeError` and crashes

**Comment by lantiga** [comment]
@hhsecond ok, thanks for bringing this up. One question (also cc @rlizzo): is there any other fundamental issue (current or future) that makes it important to not commit within the context, other than performance?

I think throwing an exception is not wrong in itself. The intention for the context is to group operations that you can then commit, so enforcing no commits reinforces that intention. @hhsecond do you see a use case where this is impractical?

One last question: @hhsecond when you say that
> It raises a `RuntimeError` and crashes
you mean that if you catch the exception we're good or it leaves the repo in an unworkable state?

BTW: we should also be able to iterate over the dataset. Maybe also having a `named_iterator` would be useful, so we're not inducing that `sample_name = str(i)`, which I don't find kosher.

**Comment by hhsecond** [comment]
So here are my thoughts.
- I don't think there are any other fundamental issues. Correct me if I am wrong @rlizzo 
- I agree throwing an exception is not wrong if that's what we decide to do. But then the error message should explicitly say it instead of `No changes made in the staging area. Cannot commit.` which is confusing
- We can definitely catch the exception but if the user knows he/she should do it, he might as well not commit in the first place. Also, once the exception raises since the write checkout was open, user will have to go and manually release the lock to continue the operations.

**Comment by rlizzo** [comment]
**So here's the good news**: There's no risk of data corruption here. When the exception is thrown, the context manager cleanup (CM) logic ensures that any data/record buffers are flushed to disk and persisted. In addition, if the user doen't try to catch the `RuntimeError` from the calling application, nothing actually happens during the commit logic - the (false-positive) error is thrown before any actual work is done - and the repository is in a fully valid state. 

HOWEVER, should the user catch that error for some reason, it's unclear what would happen. Probably nothing good, though i'm not sure. (As explained below, we can easily fix this)

This is a really nice catch @hhsecond!

### Why This Happens 

**First, no, there's nothing fundamental which prohibits making this work correctly**. Committing within a CM should work fine (though maybe frowned upon? I don't have a strong opinion here).

**Second, there actually shouldn't be a performance hit here.** Once we get the fix in, the performance would be identical to the CM method adding data followed by an `Exit` and subsequent `commit`. Let me know if I'm missing something after you read the explanation below!

The reason why this occurs in the first place is because when entering a CM, we open `lmdb transactions` and backend `file handles` for writing. The performance improvements seen are due to the records/data not actually being flushed to disk until the CM exits and pending `transactions` commit / `file handles` are closed; IO is performed in one large batch. _The design of `lmdb`, `np memmap`, and `hdf5` backends essentially only allow the writing process to see these `staged` changes until the CM exits._ (N.B. This isn't actually True in general, but it is the _effective_ behavior in the Hangar backend)

**The error occurs because the commit methods assume that all pending changes have been flushed to disk**, and that it is safe to open new `lmdb` read-optimized transaction for parsing of the user's changes. _However, since this operates in the same CM, those changes haven't been flushed to disk_, and the read `transaction` sees the `unchanged` state without any of the modifications currently buffered in memory by the write `transaction`.

The relevant code is here:

https://github.com/tensorwerk/hangar-py/blob/5ebbd993b11b40db1f06fe2b0f7d3df25852d67a/src/hangar/checkout.py#L543

So yeah, the methods thinks that there are no changes made for the simple reason that it can't actually see any! **Does this make sense?**

### The Solution

To correct this, we need to do the following:

1) Determine if the `commit` method is called from within a CM. 
2) If within a CM, manually trigger the `Exit Stack` logic from within the `commit` method (at the start of it).
3) Run the commit rest of the `commit` method logic as normal.
4) If within a CM, manually (re)open (`Enter`) the CM logic so that when control is returned to the calling application code downstream of the `commit` method, but within the same CM, any operations occur on objects with the same state they had originally. 

**While a simple fix should work just fine, this might be a good opportunity to refactor the CM logic** from manual triggers & `try-finally / flag-variable` implementations to more pythonic re-entrant CMs with proper `ExitStack callbacks` provided by the [standard library contextlib package](https://docs.python.org/3/library/contextlib.html)

### Assignee

@hhsecond, do you think you would be willing to lead the effort on this one? It's a fairly significant bug which needs a fix. I'll help in any way I can of course!

**Comment by hhsecond** [comment]
Sounds good!

**Comment by rlizzo** [comment]
Any update on this @hhsecond?

**Comment by hhsecond** [comment]
@rlizzo Not yet, will have it in my TODO soon, probably by Wednesday unless you want me to take it up sooner

**Comment by rlizzo** [comment]
That's ok. Just wanted to ping for an update here.

**Comment by rlizzo** [comment]
Fixed in #94
